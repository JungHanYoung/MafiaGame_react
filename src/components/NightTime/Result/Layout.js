import React from 'react';
import PropTypes from 'prop-types'

export default function Layout({ children, describe }) {

    return (
        <div className="game-content">
            <p className="night-subject">{describe}</p>
            {children}
        </div>
    )
}

Layout.propTypes = {
    children: PropTypes.node,
    describe: PropTypes.string.isRequired
}