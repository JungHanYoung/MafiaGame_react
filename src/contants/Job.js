export const JOB_NAME_OF_MAFIA = 'MAFIA';
export const JOB_NAME_OF_POLICE = 'POLICE';
export const JOB_NAME_OF_DOCTOR = 'DOCTOR';
export const JOB_NAME_OF_CITIZEN = 'CITIZEN';
