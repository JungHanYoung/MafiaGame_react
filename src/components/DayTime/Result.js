import React from 'react';
import PropTypes from 'prop-types';
import { useGame } from '../../context/GameContext';
import { JOB_NAME_OF_CITIZEN, JOB_NAME_OF_DOCTOR, JOB_NAME_OF_MAFIA, JOB_NAME_OF_POLICE } from '../../contants/Job';

class Result extends React.Component {
	render() {
		const { dayTimeVotedPerson, isReVoted } = this.props;
		return (
			<div>
				<h1>투표 결과</h1>
				{isReVoted ? <h2>다수결 결과가 되지 않아 재투표를 합니다.</h2>
					: <h2>{dayTimeVotedPerson.name}님이 죽으셨습니다.</h2>}
			</div>
		);
	}
}

Result.propTypes = {
	isReVoted: PropTypes.bool.isRequired,
	dayTimeVotedPerson: PropTypes.shape({
		code: PropTypes.number,
		daytimeVoted: PropTypes.number.isRequired,
		jobName: PropTypes.oneOf([JOB_NAME_OF_CITIZEN, JOB_NAME_OF_DOCTOR, JOB_NAME_OF_MAFIA, JOB_NAME_OF_POLICE])
			.isRequired,
		name: PropTypes.string.isRequired
	}).isRequired
};

export default useGame(({ state, actions }) => ({
	isReVoted: state.isReVoted,
	dayTimeVotedPerson: state.dayTimeVotedPerson
}))(Result);
