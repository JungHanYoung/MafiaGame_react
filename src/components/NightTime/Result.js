import React from 'react';
import { withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import { Object } from 'es6-shim';
import { JOB_NAME_OF_MAFIA } from '../../contants/Job';

class Result extends React.Component {

	get mafiaArray() {
		const { mafiaVotes } = this.props;

		return Object.values(mafiaVotes.toJS()).reduce(
			(prev, cur) => !!prev[cur] ? { ...prev, [cur]: prev[cur] + 1 } : { ...prev, [cur]: 1 }
			, {})
	}
	get saveSetByDoctor() {
		const { doctorVotes } = this.props;

		return new Set(Object.values(doctorVotes.toJS()))
	}

	get isRevoted() {

		const mafiaArray = this.mafiaArray

		let max = 0;
		let flag = false;
		for (let key in mafiaArray) {
			if (max < mafiaArray[key]) {
				max = mafiaArray[key]
				flag = false
			} else if (max === mafiaArray[key]) {
				flag = true
			}
		}

		return flag
	}

	get killPersonName() {
		const savePeople = this.saveSetByDoctor
		const mafiaArray = this.mafiaArray
		const killPersonName = Object.keys(mafiaArray).reduce((acc, cur) => {
			return mafiaArray[acc] < mafiaArray[cur] ? cur : acc
		})
		return savePeople.has(killPersonName)
			? null : killPersonName
	}

	get isMafiaVictory() {
		const killPersonName = this.killPersonName
		if (killPersonName) {
			const { players } = this.props
			const after = players
				.filter(player => player.get('name') !== killPersonName)

			const mafias = after.filter(player => player.get('jobName') === JOB_NAME_OF_MAFIA)
			const citizens = after.filter(player => player.get('jobName') !== JOB_NAME_OF_MAFIA)
			if (mafias.size >= citizens.size) {
				return true
			}
			// else if(mafias.size === 0) {
			// 	return 'citizen'
			// }
		}
		return false;
	}

	handleKillAndNext = () => {
		const { changeDayAndNight, deletePlayer } = this.props
		deletePlayer(this.killPersonName)
		changeDayAndNight()
	}

	render() {
		const { changeDayAndNight, voteAgain, moveToMain } = this.props
		const { killPersonName, isRevoted, isMafiaVictory } = this
		return (
			<>
				{isRevoted ? (
					<>
						<div className="game-content">
							<p className="content-description">밤 투표 결과</p>
							<h4>마피아 투표가 동률이 났습니다.</h4>
							<button className="btn-lg" onClick={changeDayAndNight}>낮으로 갑니다.</button>
							<button className="btn-lg" onClick={voteAgain}>재투표로 갑니다.</button>
						</div>
					</>
				) : isMafiaVictory ? (
					<>
						<div className="game-content">
							<p className="content-description">밤 투표 결과</p>
							<h4>마피아가 {killPersonName}님을 죽였습니다.</h4>
							<h3>마피아가 승리하였습니다.</h3>
						</div>
						<button
							className="btn-lg"
							onClick={moveToMain}>메인으로</button>
					</>

				) : <>
							<div className="game-content">
								<p className="content-description">밤 투표 결과</p>
								{killPersonName ? (
									<h4>마피아가 {killPersonName}님을 죽였습니다.</h4>
								) : (
										<h4>마피아가 시민을 죽이지 못하였습니다.</h4>
									)}
							</div>
							<button
								className="btn-lg"
								onClick={this.handleKillAndNext}>낮으로 갑니다.</button>
						</>}
			</>

		);
	}
}
// 마피아는 <누구>를 죽이고 의사는 사람을 살리지 못하였습니다.
// 마피아는 <누구>를 죽였이려 하였으나 의사가 살렸습니다.

Result.propTypes = {
	mafiaVotes: ImmutablePropTypes.map,
	doctorVotes: ImmutablePropTypes.map,
	players: ImmutablePropTypes.list,
	changeDayAndNight: PropTypes.func.isRequired,
	deletePlayer: PropTypes.func.isRequired,
	voteAgain: PropTypes.func.isRequired,
	moveToMain: PropTypes.func.isRequired
};

export default withRouter(Result)
