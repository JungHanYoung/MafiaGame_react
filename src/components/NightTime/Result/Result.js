import React from 'react';
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import { JOB_NAME_OF_MAFIA } from '../../../contants/Job';

import Layout from './Layout'

class Result extends React.Component {

    get mafiaArray() {
        const { mafiaVotes } = this.props;

        return Object.values(mafiaVotes.toJS()).reduce(
            (prev, cur) => !!prev[cur] ? { ...prev, [cur]: prev[cur] + 1 } : { ...prev, [cur]: 1 }
            , {})
    }
    get saveSetByDoctor() {
        const { doctorVotes } = this.props;

        const set = new Set(Object.values(doctorVotes.toJS()))
        // console.log('doctor Set:', set)
        return set
    }

    get isRevoted() {

        const mafiaArray = this.mafiaArray

        let max = 0;
        let flag = false;
        for (let key in mafiaArray) {
            if (max < mafiaArray[key]) {
                max = mafiaArray[key]
                flag = false
            } else if (max === mafiaArray[key]) {
                flag = true
            }
        }

        return flag
    }

    get killPersonName() {
        const savePeople = this.saveSetByDoctor
        const mafiaArray = this.mafiaArray
        const killPersonName = Object.keys(mafiaArray).reduce((acc, cur) => mafiaArray[acc] < mafiaArray[cur] ? cur : acc)
        // console.log(killPersonName);
        // console.log(savePeople.has(killPersonName))
        return savePeople.has(killPersonName)
            ? null : killPersonName
    }

    get isMafiaVictory() {
        const killPersonName = this.killPersonName
        if (killPersonName) {
            const { players } = this.props
            const after = players
                .filter(player => player.get('name') !== killPersonName)

            const mafias = after.filter(player => player.get('jobName') === JOB_NAME_OF_MAFIA)
            const citizens = after.filter(player => player.get('jobName') !== JOB_NAME_OF_MAFIA)

            if (mafias.size >= citizens.size) {
                return true
            }
        }
        return false;
    }

    handleKillAndNext = () => {
        const { changeDayAndNight, deletePlayer } = this.props
        if (this.killPersonName) {
            deletePlayer(this.killPersonName)
        }
        changeDayAndNight()
    }

    render() {
        const { changeDayAndNight, voteAgain, moveToResult } = this.props
        const { killPersonName, isRevoted, isMafiaVictory } = this
        return isRevoted ?
            (<Layout describe={`투표 결과`}>
                <p className="night-describe">마피아 투표가 동률이 났습니다.</p>
                <button className="btn-lg" onClick={changeDayAndNight}>낮으로 갑니다.</button>
                <button className="btn-lg" onClick={voteAgain}>재투표로 갑니다.</button>
            </Layout>) : isMafiaVictory ? (
                <>
                    <Layout describe={`투표 결과`}>
                        <p className="night-describe">
                            마피아가&nbsp;
							<span className="person">{killPersonName}</span>
                            &nbsp;님을 죽였습니다.
						</p>
                        <p className="night-complete">마피아가 승리하였습니다.</p>
                    </Layout>
                    <button
                        className="btn-lg"
                        onClick={moveToResult}>메인으로</button>
                </>
            ) : (
                    <>
                        <Layout describe={`투표 결과`}>
                            {killPersonName ? (
                                <h4>마피아가 {killPersonName}님을 죽였습니다.</h4>
                            ) : (
                                    <h4>마피아가 시민을 죽이지 못하였습니다.</h4>
                                )}
                        </Layout>
                        <button
                            className="btn-lg"
                            onClick={this.handleKillAndNext}>낮으로 갑니다.</button>
                    </>
                )
    }
}
// 마피아는 <누구>를 죽이고 의사는 사람을 살리지 못하였습니다.
// 마피아는 <누구>를 죽였이려 하였으나 의사가 살렸습니다.

Result.propTypes = {
    mafiaVotes: ImmutablePropTypes.map,
    doctorVotes: ImmutablePropTypes.map,
    players: ImmutablePropTypes.list,
    changeDayAndNight: PropTypes.func.isRequired,
    deletePlayer: PropTypes.func.isRequired,
    voteAgain: PropTypes.func.isRequired,
    moveToResult: PropTypes.func.isRequired
};

export default Result
