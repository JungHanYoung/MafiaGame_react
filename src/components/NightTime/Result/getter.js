import { JOB_NAME_OF_MAFIA } from "../../../contants/Job";

export const getIsMafiaVictory = (killPersonName, players) => {
    let flag = false;
    if (killPersonName) {
        const after = players
            .filter(player => player.get('name') !== killPersonName)

        const mafias = after.filter(player => player.get('jobName') === JOB_NAME_OF_MAFIA)
        const citizens = after.filter(player => player.get('jobName') !== JOB_NAME_OF_MAFIA)

        if (mafias.size >= citizens.size) {
            flag = true
        }
    }
    return flag;
}

export const getMafiaArray = (mafiaVotes) => Object.values(mafiaVotes).reduce(
    (prev, cur) => !!prev[cur] ? { ...prev, [cur]: prev[cur] + 1 } : { ...prev, [cur]: 1 }
    , {})

export const getSetSaveByDoctor = doctorVotes => new Set(Object.values(doctorVotes))

export const getIsRevoted = mafiaArray => {
    let max = 0;
    let flag = false;
    for (let key in mafiaArray) {
        if (max < mafiaArray[key]) {
            max = mafiaArray[key]
            flag = false
        } else if (max === mafiaArray[key]) {
            flag = true
        }
    }
    return flag
}

export const getKillPersonName = (mafiaArray, set) => {
    let killPersonName = Object.keys(mafiaArray).reduce((acc, cur) => mafiaArray[acc] < mafiaArray[cur] ? cur : acc)

    return set.has(killPersonName) ? null : killPersonName
}

