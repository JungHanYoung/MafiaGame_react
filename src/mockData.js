import { List, Map } from "immutable";
import { JOB_NAME_OF_CITIZEN, JOB_NAME_OF_POLICE, JOB_NAME_OF_DOCTOR, JOB_NAME_OF_MAFIA } from "./contants/Job";


export const players = List([
    Map({
        name: "한영",
        jobName: JOB_NAME_OF_CITIZEN,
        daytimeVoted: 3
    }),
    Map({
        name: "병욱",
        jobName: JOB_NAME_OF_POLICE,
        daytimeVoted: 0
    }),
    Map({
        name: "지은",
        jobName: JOB_NAME_OF_DOCTOR,
        daytimeVoted: 0
    }),
    Map({
        name: "대용",
        jobName: JOB_NAME_OF_MAFIA,
        daytimeVoted: 0
    })
])

const playerWithoutImmutable = [
    {
        name: '한영',
        jobName: JOB_NAME_OF_CITIZEN
    },
    {
        name: '병욱',
        jobName: JOB_NAME_OF_MAFIA
    },
    {
        name: '대용',
        jobName: JOB_NAME_OF_DOCTOR
    },
    {
        name: '기수',
        jobName: JOB_NAME_OF_POLICE
    },
]

export const minPropsByGame = {
    history: {
        location: {
            state: {
                players: playerWithoutImmutable
            }
        }
    }
}

export const playersWithMafia = List([
    Map({
        name: "한영",
        jobName: JOB_NAME_OF_CITIZEN,
        daytimeVoted: 0
    }),
    Map({
        name: "병욱",
        jobName: JOB_NAME_OF_POLICE,
        daytimeVoted: 0
    }),
    Map({
        name: "지은",
        jobName: JOB_NAME_OF_DOCTOR,
        daytimeVoted: 0
    }),
    Map({
        name: "대용",
        jobName: JOB_NAME_OF_MAFIA,
        daytimeVoted: 0
    }),
    Map({
        name: "민기",
        jobName: JOB_NAME_OF_MAFIA,
        daytimeVoted: 0
    }),
    Map({
        name: "하림",
        jobName: JOB_NAME_OF_MAFIA,
        daytimeVoted: 0
    }),
])