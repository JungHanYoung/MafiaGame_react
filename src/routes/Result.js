import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames'
import { JOB_NAME_OF_MAFIA } from '../contants/Job';


const propTypes = {
    history: PropTypes.object.isRequired
};

export default function Result({
    history
}) {
    const players = history.location.state.players
    return (
        <>
            <h2 className="game-title">hello mafia</h2>
            <div className="main-container">
                <div className="game-content">
                    <h3 className="result-subject">게임 결과</h3>
                    <div className="result-container">
                        <div className="scroll-wrapper">
                            {players.map(player => (
                                <div
                                    key={`result-each-player__${player.name}`}
                                    className="result-item">
                                    <div className="result-name">{player.name}</div>
                                    <div
                                        className={
                                            classNames('result-job',
                                                {
                                                    mafia: player.jobName
                                                        === JOB_NAME_OF_MAFIA
                                                })}
                                    >{player.jobName}</div>
                                </div>
                            ))}
                        </div>
                    </div>
                </div>
            </div>
            <button
                className="btn-lg"
                onClick={() => history.push('/')}>메인으로</button>
        </>
    )
}

// class Result extends Component {
//     render() {
//         const { history } = this.props
//         const players = history.location.state.players
//         return (
//             <>
//                 <h2 className="game-title">hello mafia</h2>
//                 <div className="main-container">
//                     <div className="game-content">
//                         <h3 className="result-subject">게임 결과</h3>
//                         <div className="result-container">
//                             <div className="scroll-wrapper">
//                                 {players.map(player => (
//                                     <div
//                                         key={`result-each-player__${player.name}`}
//                                         className="result-item">
//                                         <div className="result-name">{player.name}</div>
//                                         <div
//                                             className={
//                                                 classNames('result-job',
//                                                     {
//                                                         mafia: player.jobName
//                                                             === JOB_NAME_OF_MAFIA
//                                                     })}
//                                         >{player.jobName}</div>
//                                     </div>
//                                 ))}
//                             </div>
//                         </div>
//                     </div>
//                 </div>
//                 <button
//                     className="btn-lg"
//                     onClick={() => this.props.history.push('/')}>메인으로</button>
//             </>
//         );
//     }
// }

Result.propTypes = propTypes;
