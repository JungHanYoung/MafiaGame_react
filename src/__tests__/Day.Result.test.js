import React from 'react'
import { shallow, mount } from "enzyme";
import { List, Map } from 'immutable'

import Result from '../components/DayTime/Result'
import { JOB_NAME_OF_CITIZEN, JOB_NAME_OF_MAFIA, JOB_NAME_OF_DOCTOR, JOB_NAME_OF_POLICE } from '../contants/Job';
import { players } from '../mockData'
import sinon from 'sinon'

const minProps = {
    players,
    changeDayTimeOrder: jest.fn(),
    deletePlayer: jest.fn(),
    moveToResult: jest.fn(),
    moveRevote: jest.fn()
}

const playersWithMafiaVictory = List([
    Map({
        name: "대용",
        jobName: JOB_NAME_OF_MAFIA,
        daytimeVoted: 1
    }),
    Map({
        name: "한영",
        jobName: JOB_NAME_OF_CITIZEN,
        daytimeVoted: 0
    }),
    Map({
        name: "미린",
        jobName: JOB_NAME_OF_MAFIA,
        daytimeVoted: 0
    }),
    Map({
        name: "대랑",
        jobName: JOB_NAME_OF_POLICE,
        daytimeVoted: 0
    }),
    Map({
        name: "대미",
        jobName: JOB_NAME_OF_DOCTOR,
        daytimeVoted: 4
    })
])
const playersWithCitizenVictory = List([
    Map({
        name: "대용",
        jobName: JOB_NAME_OF_MAFIA,
        daytimeVoted: 3
    }),
    Map({
        name: "한영",
        jobName: JOB_NAME_OF_CITIZEN,
        daytimeVoted: 0
    }),
    Map({
        name: "대랑",
        jobName: JOB_NAME_OF_POLICE,
        daytimeVoted: 0
    }),
    Map({
        name: "대미",
        jobName: JOB_NAME_OF_DOCTOR,
        daytimeVoted: 1
    })
])



const sameRatePlayers = List([
    Map({
        name: "대용",
        jobName: JOB_NAME_OF_MAFIA,
        daytimeVoted: 0
    }),
    Map({
        name: "한영",
        jobName: JOB_NAME_OF_CITIZEN,
        daytimeVoted: 2
    }),
    Map({
        name: "미린",
        jobName: JOB_NAME_OF_DOCTOR,
        daytimeVoted: 2
    }),
    Map({
        name: "대랑",
        jobName: JOB_NAME_OF_POLICE,
        daytimeVoted: 0
    }),
])

describe('<Result /> 컴포넌트', () => {
    let wrapper = null;

    it('<Result /> 렌더링 테스트', () => {
        wrapper = mount(<Result {...minProps} />)
    })
    it('props 받기', () => {
        // props > players 가 잘 받아지는가. (플레이어 4명)
        expect(wrapper.prop('players').size).toEqual(4)
    })
    it('기본 props에 대한 렌더링 테스트', () => {

        expect(wrapper.find('div.game-content').exists()).toEqual(true)
        expect(wrapper.find('p.content-description').exists()).toEqual(true)
        expect(wrapper.find('p.content-description').text()).toEqual('한영 님이 죽으셨습니다.')
        expect(wrapper.find('button.btn-lg').exists()).toEqual(true)
        expect(wrapper.find('button.btn-lg').text()).toEqual('밤이 됩니다.')
    })

    it('동률의 투표가 발생했을 경우의 렌더링 테스트', () => {
        const deletePlayer = sinon.spy()
        // players props에 동률 투표의 상태를 적용
        wrapper = shallow(<Result {...minProps} players={sameRatePlayers} deletePlayer={deletePlayer} />)
        expect(wrapper.find('p.content-description').exists()).toEqual(true)
        // '투표가 동률이 났습니다.'라는 텍스트가 출력되는지 확인
        expect(wrapper.find('p.content-description').text()).toEqual('투표가 동률이 났습니다.')

        expect(wrapper.find('button.btn-lg')).toHaveLength(2)

        expect(wrapper.find('button.btn-lg')).toMatchSnapshot()

        const revoteBtn = wrapper.find('button.btn-lg').at(0)
        const setNightBtn = wrapper.find('button.btn-lg').at(1)

        setNightBtn.simulate('click')

        expect(deletePlayer.calledOnce).not.toBe(1)

        revoteBtn.simulate('click')

        expect()
    })

})

describe('마피아 승리 조건일때 테스트', () => {
    const moveToResult = sinon.spy()
    const wrapper = shallow(<Result {...minProps} players={playersWithMafiaVictory} moveToResult={moveToResult} />)

    it('마피아 승리 메세지 출력', () => {
        const victoryMessage = wrapper.find('h3')

        expect(victoryMessage.exists()).toEqual(true)
    })

    it('click 이벤트에 대한 메소드 호출', () => {
        const button = wrapper.find('button.btn-lg');
        button.simulate('click')
        expect(moveToResult.callCount).toBe(1)
    })
})

describe('시민 승리 조건일때 테스트', () => {
    const moveToResult = sinon.spy()
    const wrapper = shallow(<Result {...minProps} players={playersWithCitizenVictory} moveToResult={moveToResult} />)

    it('시민 승리 메세지 출력', () => {
        const message = wrapper.find('h3')
        expect(message.text()).toBe('시민이 승리하였습니다.')
    })

    it('버튼 클릭 이벤트에 대한 메소드 호출', () => {
        const button = wrapper.find('button.btn-lg');
        button.simulate('click')
        expect(moveToResult.callCount).toBe(1)
    })
})