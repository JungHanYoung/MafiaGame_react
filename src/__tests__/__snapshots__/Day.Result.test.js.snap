// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Result /> 컴포넌트 동률의 투표가 발생했을 경우의 렌더링 테스트 1`] = `
Array [
  <button
    className="btn-lg"
    onClick={[MockFunction]}
  >
    재투표
  </button>,
  <button
    className="btn-lg"
    onClick={[Function]}
  >
    밤이됩니다.
  </button>,
]
`;
